src/local-mode.ts

```diff
@@ -56,42 +90,172 @@ export class LocalModeManager {
 
 		if (!localMode) {
 			// Remove any leftovers if necessary
-			await this.removeLocalModeArtifacts();
+			await this.handleLocalModeStateChange(false);
 		}
 	}
 
-	public async removeLocalModeArtifacts(): Promise<void> {
-		try {
-			const images = await this.getLocalModeImages();
-			const containers = await this.getLocalModeContainers(images);
+	// Query the engine to get currently running containers and installed images.
+	public async collectEngineSnapshot(): Promise<EngineSnapshotRecord> {
+		const containers = this.docker
+			.listContainers()
+			.then(resp => _.map(resp, 'Id'));
+		const images = this.docker.listImages().then(resp => _.map(resp, 'Id'));
+		const volumes = this.docker
+			.listVolumes()
+			.then(resp => _.map(resp.Volumes, 'Name'));
+		const networks = this.docker.listNetworks().then(resp => _.map(resp, 'Id'));
+
+		return Bluebird.all([containers, images, volumes, networks]).then(
+			data =>
+				new EngineSnapshotRecord(
+					new EngineSnapshot(data[0], data[1], data[2], data[3]),
+					new Date(),
+				),
+		);
+	}
+
+	// Store engine snapshot data in the local database.
+	public async storeEngineSnapshot(
+		record: EngineSnapshotRecord,
+	): Promise<void> {
+		const timestamp = record.timestamp.toISOString();
+		log.info(
+			`Storing engine snapshot in the database. Timestamp: ${timestamp}`,
+		);
+		return this.db
+			.models('engineSnapshot')
+			.delete()
+			.then(() => {
+				return this.db.models('engineSnapshot').insert({
+					snapshot: JSON.stringify(record.snapshot),
+					timestamp,
+				});
+			});
+	}
 
-			await Bluebird.map(containers, containerId => {
-				log.debug('Removing local mode container: ', containerId);
-				return this.docker.getContainer(containerId).remove({ force: true });
+	// Read the latest stored snapshot from the database.
+	public async retrieveLatestSnapshot(): Promise<EngineSnapshotRecord | null> {
+		return this.db
+			.models('engineSnapshot')
+			.select()
+			.orderBy('rowid', 'DESC')
+			.first()
+			.then(r => {
+				if (!r) {
+					return null;
+				}
+				return new EngineSnapshotRecord(
+					EngineSnapshot.fromJSON(r.snapshot),
+					new Date(Date.parse(r.timestamp)),
+				);
 			});
-			await Bluebird.map(images, imageId => {
-				log.debug('Removing local mode image: ', imageId);
-				return this.docker.getImage(imageId).remove({ force: true });
+	}
+
+	private async removeLocalModeArtifacts(
+		objects: EngineSnapshot,
+	): Promise<void> {
+		log.debug(`Going to delete following objects: ${objects}`);
```

Comment:
This is probably going to need to be formatted a bit nicer, can you paste what it currently prints please?

Reply:
```
Going to delete the following objects: containers [container-1,container-2], images [image-1,image-2], volumes [volume-1,volume-2], networks [network-1,network-2]
```

EngineSnapshot has toString method. But for some reason it's not committed in this draft... I will push soon.

Reply:
See https://github.com/balena-io/balena-supervisor/pull/1012/files#diff-0f5fedf56e0f6f78a6c52a828857e5e4R40